tests/ui/lint/dead-code/lint-dead-code-4.rs

```diff
@@ -12,10 +12,10 @@ fn field_read(f: Foo) -> usize {
 }
 
 enum XYZ {
-    X, //~ ERROR variants `X` and `Y` are never constructed
+    X, //~ ERROR variant `X` is never constructed
```

Comment:
Hmm, I'm not sure about this change. Because the old statement that `Y` is never constructed is correct. And if a variant appears in a pattern but is never constructed I think it should be warned about, because that could genuinely indicate dead code.

So now I'm questioning the whole premise of #132874 and this PR. If anything, I think `let Foo(_x) = get_thing::<Foo>();` should be made to give a warning. Does that seem reasonable?

Reply:
Also, it would be better if the commits were squashed together because they are all part of the same logical change. (But that may be moot if the PR doesn't go ahead.)

Reply:
(Also, you've done a nice job on the code and test changes, they look good if we are willing to go in this direction.)

Reply:
Thank you for your review. I understood that the code given in the issue #132874 should be warned as you explained.

Now, I'm thinking that there's a possibility that the issue owner actually found a bug, but the MRE in the issue was over-generalized. So, I'm asking to give us another MRE closer to the original code.
https://github.com/rust-lang/rust/issues/132874#issuecomment-2495781342

Reply:
@nnethercote 
The issue author gave us another repro as follows, which is closer to the real code.
https://github.com/rust-lang/rust/issues/132874#issuecomment-2496263994

The new code indicates there are cases where a struct is actually used without being constructed in the normal way. I think this code should be compiled without any warning.

Though the best is to achieve both allowing the above code to be compiled without any warning and keeping the warning to the cases like `tests/ui/lint/dead-code/lint-dead-code-4.rs`, I'm not sure whether it can be achieved.

So, can I have your advice on which direction to go?

Reply:
That is a tricky case: a `PacketHeader` is constructed, but via binary deserialization, which the compiler cannot understand. It seems like a false positive to avoid without introducing false negatives in other cases. I'm not sure how best to handle it. I have asked for other opinions in [this Zulip thread](https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/Tricky.20.22never.20constructed.22.20case).